From b7c96cf56398cecf74c7aacc7845e9763f1e76b1 Mon Sep 17 00:00:00 2001
From: "Kevin F. Haggerty" <haggertk@lineageos.org>
Date: Sat, 25 Aug 2018 15:00:25 -0600
Subject: [PATCH 9/9] libui: Allow extension of valid gralloc 1.0 buffer usage
 bits

* Change I7f4174581e24e361577640b9263514a168ed482d implemented
  validation of the buffer description info prior to creating
  the descriptor. Some of our legacy devices need to whitelist
  additional usage bits to support various functionality.

Change-Id: Ie369e78f78e9ac0b18ab3dfea520d4f123005d92
---
 libs/ui/Android.bp   | 8 ++++++++
 libs/ui/Gralloc2.cpp | 6 ++++++
 2 files changed, 14 insertions(+)

diff --git a/libs/ui/Android.bp b/libs/ui/Android.bp
index ff9d19e29..25c353a51 100644
--- a/libs/ui/Android.bp
+++ b/libs/ui/Android.bp
@@ -46,6 +46,14 @@ cc_library_shared {
         "-Wno-padded",
     ],
 
+    product_variables: {
+        lineage: {
+            additional_gralloc_10_usage_bits: {
+                cppflags: ["-DADDNL_GRALLOC_10_USAGE_BITS=%s"],
+            },
+        },
+    },
+
     sanitize: {
         integer_overflow: true,
     },
diff --git a/libs/ui/Gralloc2.cpp b/libs/ui/Gralloc2.cpp
index b92cbf351..918251f9d 100644
--- a/libs/ui/Gralloc2.cpp
+++ b/libs/ui/Gralloc2.cpp
@@ -45,6 +45,12 @@ uint64_t getValid10UsageBits() {
         // TODO(b/72323293, b/72703005): Remove these additional bits
         bits = bits | (1 << 10) | (1 << 13);
 
+#ifdef ADDNL_GRALLOC_10_USAGE_BITS
+        uint64_t addnl_bits = static_cast<uint64_t>(ADDNL_GRALLOC_10_USAGE_BITS);
+        ALOGI("Adding additional valid usage bits: 0x%" PRIx64, addnl_bits);
+        bits = bits | addnl_bits;
+#endif
+
         return bits;
     }();
     return valid10UsageBits;
-- 
2.17.1

